Add documentation for virtual-outbound (#508)
kumahq/kuma-website: docs/docs/1.2.3/policies/virtual-outbound.md

# Virtual Outbounds

This policy enables users to create ways to reach a specific set of dataplanes with a `hostname:port` generated from the dataplanes tags.

Possible use cases are:

1) Preserving hostnames when migrating to service mesh (This enables migration without modifying app configuration).
2) Providing multiple hostnames for reaching the same service (when renaming or for usability).
3) Providing specific routes (reach a specific pod in a service when using StatefulSets, add an url to reach a specific version of a service).
4) Expose multiple inbounds on different ports.

> 

Reviewer: ```suggestion

Note that complex virtual outbounds do not work for cross-zone traffic. This is because only service tags are propagated across zones. 
```
(moved up from old Collisions section)

    - name: service
      tagKey: "kuma.io/service"
    - name: port
      tagKey: port
    - name: v
      tagKey: version
```

Will create a hostname: `v1.backend.mesh` and port: `1800`.

Virtual Outbounds have some constraints/limitations:

- It only works when using transparent proxy.
- When not using the [data plane proxy DNS](/docs/1.2.3/networking/dns.md#data-plane-proxy-dns), all generated hostnames must end with the value of the configuration `dns_server.domain` (whose default is `.mesh`).
- If a `tagKey` is absent it uses `name` as `tagKey`.
- `name` must be alphanumeric as it's used as a go template key.
- for a virtual-outbound set of parameters all `name` must be unique.
- `kuma.io/service` must be in the parameters even if it's unused in the template (this prevents defining hostnames that spans services).

For each virtual outbound the Kuma control plane will process all dataplanes matching the selector.
It will then apply the templates for `conf.host` and `conf.port` and assign a virtual ip for each unique set defined by all the `tagKeys` value in the parameters.
This means that different hostnames may resolve to the same vip because they map to the same tag set.

> ### Collisions

Reviewer: ```suggestion
## Collisions
```
I called everything "duplicates" in suggestions further on. Is it appropriate as the heading also? Wrong in the details? If it _is_ appropriate, suggest renaming this section to something like `## Duplicate hostnames`

H2 lets it show up in the left nav (more similar suggestions follow). Better not to nest too deep in general.

And let's move this whole section to the bottom of the page.

Author: I'd prefer to stay with collisions. I've picked collisions because that's how people name these things: `hash collisions`, `name collisions`, `address collisions`.

> We rely on user defined templates it is thus possible to have hostname collisions.

Reviewer: I think we can remove this sentence -- it's an implementation detail that seems sufficiently obvious? Or am I missing something?

> We apply these rules to avoid them:
> 
> 1) For collisions between virtual outbounds, the virtual outbound with the highest priority takes over as [explained in the policy doc](../../1.2.2/policies/how-kuma-chooses-the-right-policy-to-apply.md).
> 2) For all collisions log messages will be emitted. 
> 3) Because only service tags are propagated across zones complex virtual outbounds will not work for cross-zone traffic. 

Reviewer: ```suggestion
When duplicate hostnames are detected, the virtual outbound with the highest priority takes over. For more information, see [the documentation on ow Kuma chooses the right poliy](../../1.2.2/policies/how-kuma-chooses-the-right-policy-to-apply.md). All duplicate instances are logged. 
```

n.b content removed from here is put further up the page bc it's better to provide that kind of warning early in the doc :D

> ### Examples

Reviewer: ```suggestion
## Usage

The following examples show how to use virtual outbounds for different use cases.
```

> #### Same as the default DNS

Reviewer: ```suggestion
### Same as the default DNS
```

:::: tabs :options="{ useUrlFragment: false }"
::: tab "Kubernetes"
```yaml
apiVersion: kuma.io/v1alpha1
kind: Mesh
type: VirtualOutbound
mesh: default
name: default
selectors:
  - match:
      kuma.io/service: "*"
conf:
  host: "{{.service}}.mesh"
  port: 80
  parameters:
    - name: service
      tagKey: "kuma.io/service"
```
:::
::: tab "Universal"
```yaml
type: VirtualOutbound
mesh: default
name: default
selectors:
  - match:
      kuma.io/service: "*"
conf:
  host: "{{.service}}.mesh"
  port: 80
  parameters:
    - name: service
      tagKey: "kuma.io/service"
```
:::
::::

> #### One hostname per version

Reviewer: ```suggestion
### One hostname per version
```

:::: tabs :options="{ useUrlFragment: false }"
::: tab "Kubernetes"
```yaml
apiVersion: kuma.io/v1alpha1
kind: Mesh
type: VirtualOutbound
mesh: default
name: default
selectors:
  - match:
      kuma.io/service: "*"
conf:
  host: "{{.service}}.{{.version}}.mesh"
  port: 80
  parameters:
    - name: service
      tagKey: "kuma.io/service"
    - name: version
      tagKey: "kuma.io/version"
```
:::
::: tab "Universal"
```yaml
type: VirtualOutbound
mesh: default
name: default
selectors:
  - match:
      kuma.io/service: "*"
conf:
  host: "{{.service}}.{{.version}}.mesh"
  port: 80
  parameters:
    - name: service
      tagKey: "kuma.io/service"
    - name: version
      tagKey: "kuma.io/version"
```
:::
::::

> #### Custom tag to define the hostname and port

Reviewer: ```suggestion
### Custom tag to define the hostname and port
```

:::: tabs :options="{ useUrlFragment: false }"
::: tab "Kubernetes"
```yaml
apiVersion: kuma.io/v1alpha1
kind: Mesh
type: VirtualOutbound
mesh: default
name: default
selectors:
  - match:
      kuma.io/service: "*"
conf:
  host: "{{.hostname}}"
  port: "{{.port}}"
  parameters:
    - name: hostname
      tagKey: "my.mesh/hostname"
    - name: port
      tagKey: "my.mesh/port"
```
:::
::: tab "Universal"
```yaml
type: VirtualOutbound
mesh: default
name: default
selectors:
  - match:
      kuma.io/service: "*"
conf:
  host: "{{.hostname}}"
  port: "{{.port}}"
  parameters:
    - name: hostname
      tagKey: "my.mesh/hostname"
    - name: port
      tagKey: "my.mesh/port"
    - name: service
```
:::
::::

> #### One hostname per instance

Reviewer: ```suggestion
### One hostname per instance
```

This enables reaching specific dataplanes in a service.

> This is especially useful for running distributed databases like Kafka, Zookeeper...

Reviewer: ```suggestion
This is especially useful for running distributed databases such as Kafka or Zookeeper.
```
